File: examples/order-matching/web2/csharp/order-matching-api/Controllers/OrderMatchingController.cs
using Microsoft.AspNetCore.Mvc;

namespace order_matching_api.Controllers;

[ApiController]
[Route("api/orders")]
public class OrderMatchingController : ControllerBase
{
    public class Order { public string Trader { get; set; } = ""; public string Side { get; set; } = ""; public double Price { get; set; } public int Quantity { get; set; } }
    
    private static readonly List<Order> _bids = new();
    private static readonly List<Order> _asks = new();
    private static readonly object _lock = new();

    [HttpPost]
    public IActionResult PlaceOrder([FromBody] Order order)
    {
        lock (_lock)
        {
            int remaining = order.Quantity;
            if (order.Side == "BUY")
            {
                remaining = Match(order, _asks, true);
                if (remaining > 0) {
                    order.Quantity = remaining;
                    _bids.add(order);
                    _bids.Sort((a,b) => b.Price.CompareTo(a.Price));
                }
            }
            else
            {
                remaining = Match(order, _bids, false);
                if (remaining > 0) {
                    order.Quantity = remaining;
                    _asks.add(order);
                    _asks.Sort((a,b) => a.Price.CompareTo(b.Price));
                }
            }
            return Ok(new { bids = _bids, asks = _asks });
        }
    }

    private int Match(Order o, List<Order> opposites, bool isBuy)
    {
        int rem = o.Quantity;
        for (int i = 0; i < opposites.Count && rem > 0; )
        {
            var best = opposites[i];
            bool canMatch = isBuy ? (o.Price >= best.Price) : (o.Price <= best.Price);
            if (!canMatch) break;

            int qty = Math.Min(rem, best.Quantity);
            rem -= qty;
            best.Quantity -= qty;
            if (best.Quantity == 0) opposites.RemoveAt(i);
            else i++;
        }
        return rem;
    }
}

File: examples/order-matching/web2/springboot/src/main/java/com/example/demo/OrderMatchingController.java
package com.example.demo;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.ResponseEntity;
import java.util.*;

@RestController
@RequestMapping("/api/orders")
public class OrderMatchingController {

    public static class Order {
        public String trader;
        public String side;
        public double price;
        public int quantity;
    }

    private final List<Order> bids = new ArrayList<>();
    private final List<Order> asks = new ArrayList<>();

    @PostMapping
    public synchronized ResponseEntity<?> placeOrder(@RequestBody Order order) {
        int remaining = order.quantity;

        if (order.side.equals("BUY")) {
            remaining = match(order, asks, true);
            if (remaining > 0) {
                order.quantity = remaining;
                bids.add(order);
                bids.sort((a, b) -> Double.compare(b.price, a.price));
            }
        } else {
            remaining = match(order, bids, false);
            if (remaining > 0) {
                order.quantity = remaining;
                asks.add(order);
                asks.sort(Comparator.comparingDouble(a -> a.price));
            }
        }

        return ResponseEntity.ok(Map.of("bids", bids, "asks", asks));
    }

    private int match(Order newOrder, List<Order> opposites, boolean isBuy) {
        Iterator<Order> it = opposites.iterator();
        int rem = newOrder.quantity;
        while (it.hasNext() && rem > 0) {
            Order best = it.next();
            boolean canMatch = isBuy ? (newOrder.price >= best.price) : (newOrder.price <= best.price);
            if (!canMatch) break;

            int matchQty = Math.min(rem, best.quantity);
            rem -= matchQty;
            best.quantity -= matchQty;
            if (best.quantity == 0) it.remove();
        }
        return rem;
    }
}
